### PR TITLE
Add support for remapping edge prediction results.

### PR DESCRIPTION
**Pull request by classicsong**
Add graphstrom.gconstruct.remap_result.py to remap edge prediction results.
Will add support for remapping node prediction results and embeddings.

#508 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gconstruct/id_map.py [hunk 1]
```diff
@@ -66,6 +70,64 @@ def save(self, file_path):
...
+        tensor: A tensor of raw IDs.
+        """
+        if len(ids) == 0:
+            return np.array([], dtype=np.str)
+
+        return self._ids[ids]
+
+    @property
+    def size(self):
+        "Get the approximated szie of the id mapping"
```
Comment (zhjwy9343) [inline python/graphstorm/gconstruct/id_map.py:128]: "szie" -> "size".

Comment (classicsong) [reply python/graphstorm/gconstruct/id_map.py:128]: We do not need it now.

Comment (classicsong) [reply python/graphstorm/gconstruct/id_map.py:128]: I will remove it.

File: tests/end2end-tests/graphstorm-ec/mgpu_test.sh [hunk 2]
```diff
@@ -129,7 +129,7 @@ then
 fi
 
 cnt=$(ls -l /data/gsgnn_ec/prediction/user_rating_movie/ | grep predict | wc -l)
-if test $cnt != $NUM_INFO_TRAINERS
+if test $cnt != $NUM_INFO_TRAINERS * 2
```
Comment (zheng-da) [inline tests/end2end-tests/graphstorm-ec/mgpu_test.sh:132]: why do we have this change?

Comment (classicsong) [reply tests/end2end-tests/graphstorm-ec/mgpu_test.sh:132]: If we set --preserve-input True, then we have both old prediction result and new prediction result under /data/gsgnn_ec/prediction/user_rating_movie/

File: python/graphstorm/gconstruct/remap_result.py [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+    """ Do one edge remapping task
+
+        Parameters
+        ----------
+        pred_file_path: str
+            The path to the prediction result.
+        src_nid_path: str
+            The path to the file storing source node ids
+        dst_nid_path: str
+            The path to the file storing destination node ids
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/remap_result.py:58]: should we store them in separate files?

File: python/graphstorm/gconstruct/remap_result.py [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+        chunk_size: int
+            Max number of raws per output file.
+        preserve_input: bool
+            Whether the input data should be removed.
+    """
+    pred_result = th.load(pred_file_path).numpy()
+    src_nids = th.load(src_nid_path).numpy()
+    dst_nids = th.load(dst_nid_path).numpy()
+    src_id_map = id_maps[src_type]
+    dst_id_map = id_maps[dst_type]
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/remap_result.py:74]: id_maps is a global variable?

Comment (zheng-da) [reply python/graphstorm/gconstruct/remap_result.py:74]: if so, why not make it a class?

File: python/graphstorm/gconstruct/remap_result.py [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+        data = {"pred": pred,
+                "src_nid": src_nid,
+                "dst_nid": dst_nid}
+
+        write_data_parquet(data, output_fname)
+
+    if preserve_input is False:
+        os.remove(pred_file_path)
+        os.remove(src_nid_path)
+        os.remove(dst_nid_path)
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/remap_result.py:93]: do we need to remove the files directly?

File: python/graphstorm/gconstruct/remap_result.py [hunk 6]
```diff
@@ -0,0 +1,477 @@
...
+    try:
+        while True:
+            # If the queue is empty, it will raise the Empty exception.
+            idx, task_args = task_queue.get_nowait()
+            logging.debug("worker %d Processing %s task", worker_id, idx)
+            func(**task_args)
+    except queue.Empty:
+        pass
+
+def multiprocessing_remap(tasks, num_proc, remap_func):
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/remap_result.py:116]: should we make this a standardized utility function for multiprocessing?

File: python/graphstorm/gconstruct/remap_result.py [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+        #    src_nids-00000.pt
+        #    src_nids-00001.pt
+        #    ...
+        #    dst_nids-00000.pt
+        #    dst_nids-00001.pt
+        #    ...
+
+        The output emb files will be
+        #    predict-00000_00000.parquet
+        #    predict-00000_00001.parquet
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/remap_result.py:207]: why the remapped data are stored in a different format?

File: python/graphstorm/gconstruct/remap_result.py [hunk 8]
```diff
@@ -0,0 +1,477 @@
...
+        out_pred_dir = os.path.join(output_dir, "_".join(etype))
+        etype_pred_files = os.listdir(input_pred_dir)
+        src_nid_files = [fname for fname in etype_pred_files if fname.startswith("src_nids")]
+        dst_nid_files = [fname for fname in etype_pred_files if fname.startswith("dst_nids")]
+        pred_files = [fname for fname in etype_pred_files if fname.startswith("predict")]
+        src_nid_files.sort()
+        dst_nid_files.sort()
+        pred_files.sort()
+
+        num_parts = len(pred_files)
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/remap_result.py:244]: should you verify if src_nid_files and dst_nid_files have the same number of files?

File: python/graphstorm/gconstruct/remap_result.py [hunk 9]
```diff
@@ -0,0 +1,477 @@
...
+                       help="The logging level. The possible values: debug, info, warning, \
+                                   error. The default value is info.")
+
+    group.add_argument("--pred-etypes", type=str, nargs="+", default=None,
+                       help="[Optional] A list of canonical edge types which have"
+                                "prediction results For example, "
+                                "--pred-etypes user,rate,movie user,watch,movie"
+                                "If pred_etypes is not provided, result_info.json "
+                                "under prediction_dir will be used to retrive the pred_etypes")
+    group.add_argument("--output-chunk-size", type=int, default=100000,
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/remap_result.py:466]: why do we want to output the remapped results in a different number of files?

File: python/graphstorm/run/launch.py [hunk 10]
```diff
@@ -323,6 +323,58 @@ def construct_torch_dist_launcher_cmd(
...
+
+    udf_command[0] = "graphstorm.gconstruct.remap_result"
+
+    # Add remap related arguments
+    udf_command += ["--rank", str(rank)]
+    udf_command += ["--world-size", str(world_size)]
+    udf_command += ["--with-shared-fs", "True" if with_shared_fs else "False"]
+    udf_command += ["--num-processes", str(num_trainers)]
+    udf_command += ["--output-chunk-size", str(output_chunk_size)]
+    udf_command += ["--preserve-input", "True" if preserve_input else "False"]
```
Comment (zheng-da) [inline python/graphstorm/run/launch.py:367]: why do we give users this option?

File: python/graphstorm/gconstruct/remap_result.py [hunk 11]
```diff
@@ -0,0 +1,477 @@
...
+        src_nid_files = [fname for fname in etype_pred_files if fname.startswith("src_nids")]
+        dst_nid_files = [fname for fname in etype_pred_files if fname.startswith("dst_nids")]
+        pred_files = [fname for fname in etype_pred_files if fname.startswith("predict")]
+        src_nid_files.sort()
+        dst_nid_files.sort()
+        pred_files.sort()
+
+        num_parts = len(pred_files)
+        logging.debug("%s has %d embedding files", etype, num_parts)
+        if with_shared_fs:
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/remap_result.py:246]: i suppose only this option has been tested.

Comment (zheng-da) [reply python/graphstorm/gconstruct/remap_result.py:246]: did you test the other option?
if not, maybe we can remove that option first.

File: python/graphstorm/gconstruct/remap_result.py (hunk 3 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:58]: We already store them in separate files. src_nids_xxx and dst_nids_xxx.

File: python/graphstorm/gconstruct/remap_result.py (hunk 4 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:74]: It is a global variable. The purpose is to share it across forked processes with out passing id_maps to other processes (using the copy-on-write mechanism provided by Linux).

File: python/graphstorm/gconstruct/remap_result.py (hunk 5 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:93]: By default we can remove them. Users only need the remapped data.

File: python/graphstorm/gconstruct/remap_result.py (hunk 6 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:116]: You mean we move it into utils.py?
In the following PRs, both remap_node_predictions and remap_node_embeddings will use this function.

File: python/graphstorm/gconstruct/remap_result.py (hunk 7 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:207]: We can hardly store strings in torch tensor.
We supports parquet now. But we can extend it into csv/json or other formats if needed.

File: python/graphstorm/gconstruct/remap_result.py (hunk 8 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:244]: Good point

File: python/graphstorm/run/launch.py (hunk 10 above)
Comment (classicsong) [reply python/graphstorm/run/launch.py:367]: This is for debug purpose, in case the remap function has a bug.

Comment (classicsong) [reply python/graphstorm/run/launch.py:367]: By default, it is False.

File: python/graphstorm/gconstruct/remap_result.py (hunk 11 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:246]: In following PRs, I will add sagemaker support. with_shared_fs == False will be tested.

Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:246]: BTW, let me see if I can write some unitest for it.

Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:246]: Add a test in data_process/test.sh

File: python/graphstorm/gconstruct/remap_result.py (hunk 9 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:466]: Currently the number of prediction result or embedding files are same as the number of trainers. 
When the graph is large, it is possible that a single output file can be tens of GBs or even larger.
With this parameter, we can split those files into smaller files. For example thousands of parquet files, each of which takes only hundreds of MB space.

File: python/graphstorm/gconstruct/remap_result.py (hunk 4 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/remap_result.py:74]: please write comments to explain this.

File: python/graphstorm/gconstruct/remap_result.py (hunk 5 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/remap_result.py:93]: these are intermediate files for command line tools, but they are not for customer model API. the users of customer model API and the users of command line tools have very different user experience.

File: python/graphstorm/gconstruct/remap_result.py (hunk 6 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/remap_result.py:116]: i think so. we might want to generalize it for other purpose as well. for example, gconstruct has a multi-processing function as well.

File: python/graphstorm/gconstruct/remap_result.py (hunk 7 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/remap_result.py:207]: why can we not store them in torch tensors?

File: python/graphstorm/gconstruct/remap_result.py (hunk 9 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/remap_result.py:475]: i see. the default behavior should be not splitting the files.

File: python/graphstorm/gconstruct/remap_result.py (hunk 7 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:207]: Does torch support string tensors?
I think parquet is a general format that can be easily accepted by many downstream applications.

File: python/graphstorm/gconstruct/remap_result.py (hunk 5 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:93]: If users do not call remap, those files are not removed.

File: python/graphstorm/gconstruct/remap_result.py (hunk 6 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:116]: OK. I moved it into utils.py as named it multiprocessing_exec_no_return.

**Pull request merged**